Update interface and `README.md` (#120)
tweag/pirouette: README.md

_Pirouette is a research prototype under active development_

* [Use case: Example](#example-language)
* [Use case: Plutus Smart Contracts](#plutus-smart-contracts)
* [Limitations](#limitations)
* [Building, Installing and Hacking](#building-installing-and-hacking)
* [Contributing](#contributing)
* [License](#license)

> Pirouette is being built as framework for constructing different static

Author: Agree!

> analisys tools for [System F][systemf] based languages. It has a number of

Author: I'll spell check before merging

transformations (monomorphization, defunctionalization, prenex, etc...) defined
and relies on the `-XQuasiQuotes` extension to be able to easily write

> properties in the target language.

Author: Agree, I moved it further down

> There is one analisys currently built into pirouette, which symbolic evaluates terms up to

Reviewer: ```suggestion
There is one analysis currently built into pirouette, which symbolically evaluates terms up to
```

a certain bound while looking for counterexamples based on [incorrectness][incorrectness]

> or Hoare triples.

Author: Fair! I expanded it a little in a separate section.

[systemf]: https://en.wikipedia.org/wiki/System_F
[incorrectness]: https://dl.acm.org/doi/pdf/10.1145/3371078
[tweag-blogpost]: TODO

# Use Cases

## Example Language

The `Language.Pirouette.Example` defines the `Ex` language: pirouette's example language.
We refer the interested reader to check the respective module for guidelines on using pirouette over
your own language.

Now, say we have the simple program:

```haskell
addone :: Integer -> Integer
addone x = x + 1
```

And we wish to check the incorrectness triple `[x > 0] addone x [result > 0]`, that
is: if `addone x > 0`, then `x > 0`. We obviously expect a counterexample with `x = 0`.
We can use pirouette to find that assignment:

```haskell

> parms :: IncorrectnessParams Ex
> parms = IncorrectnessParams

Reviewer: ```suggestion
params :: IncorrectnessParams Ex
params = IncorrectnessParams
```

>           [term| \x:Integer . x + 1 |] -- Term to symbolically evaluate
>           [ty| Integer |] -- type of the term above
>           -- Conditions to check
>           ([term| \(res:Integer) (x:Integer) . 0 < res |]
>             :==>: [term| \(res:Integer) (x:Integer) . 0 < x |])

Reviewer 2: ```suggestion
          [term| \x:Integer . x + 1 |] -- A function term to symbolically evaluate
          [ty| Integer |] -- type of the result of the function term above
          -- Conditions to check, they take the result and the inputs of the function term 
          ([term| \(res:Integer) (x:Integer) . 0 < res |]
            :==>: [term| \(res:Integer) (x:Integer) . 0 < x |])
```

```

Now we can run `replIncorrectnessLogic1 10 params` in a repl and we should see:
```
💸 COUNTEREXAMPLE FOUND
{ __result ↦ 1
  x ↦ 0 }
```

## Plutus Smart Contracts

[Plutus] is a subset of Haskell used to
write smart contracts for Cardano blockchain, which utilizes the _Extended UTxO_
[[1](https://iohk.io/en/research/library/papers/the-extended-utxo-model/),
[2](https://iohk.io/en/research/library/papers/native-custom-tokens-in-the-extended-utxo-model/)]

> model to represent its accounts. Pirouette has a prototype interface to interacting with

Reviewer 3: I'm not sure the EUTxO part is worth mentioning here in the context of Pirouette.

The folks who know about Plutus likely know this, and the folks who don't probably don't care.

plutus through `Language.Pirouette.PlutusIR`, but further work is necessary to
seamlessly interact with validator scripts.

An example of using pirouette to load and check an incorrectness triple over a `.pir` file

> can be found in [here](tests/unit/Language/Pirouette/PlutusIR/SymEvalSpec.hs).

Reviewer: ```suggestion
can be found [here](tests/unit/Language/Pirouette/PlutusIR/SymEvalSpec.hs).
```

You might want to consider `direnv` and [`nix-direnv`](https://github.com/nix-community/nix-direnv)
instead of running `nix-shell` directly.

## Limitations

In its current form, `pirouette` itself is still a research prototype
and, hence, it has plenty of limitations. The overall approach of pirouette's

> symbolic engine is an intersetion of bounded model checking and symbolic executon, hence,

Reviewer: ```suggestion
symbolic engine is an intersection of bounded model checking and symbolic execution, hence,
```

> it also carries the limitations of a model checker: it can be slow
> space can be slow and a model checker is not a substitute for a formal

Author: ```suggestion
it also carries the limitations of a model checker: exploration of the state space can be slow
and it is not a substitute for a formal
```

proof of correctness.

> 

Author: Indeed, it was a copy paste error. I fixed it, thanks!

You might want to consider `direnv` and [`nix-direnv`](https://github.com/nix-community/nix-direnv)
instead of running `nix-shell` directly.

## Limitations

> In its current form, `pirouette` itself is still a research prototype
> and, hence, it has plenty of limitations. The overall approach of pirouette's

Reviewer 3: ```suggestion
In its current form, `pirouette` itself is still a research prototype,
hence, it has plenty of limitations. The overall approach of pirouette's
```

symbolic engine is an intersection of bounded model checking and symbolic execution, hence,
it also carries the limitations of a model checker: exploration of the state space can be slow

> and it is not a substitute for a formal

Reviewer 3: This is indeed very good to mention!
